optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/common/accessor/gizmo/GizmoMemberDescriptor.java

```diff
@@ -239,7 +282,31 @@ public String getTypeName() {
             holder[0] = md.getReturnType();
         });
 
-        return Type.getType(holder[0]).getClassName();
+        return org.objectweb.asm.Type.getType(holder[0]).getClassName();
+    }
+
+    public Type getType() {
+        Type[] out = new Type[1];
+
+        whenMetadataIsOnField(fd -> {
+            try {
+                out[0] = declaringClass.getDeclaredField(fd.getName()).getGenericType();
+            } catch (NoSuchFieldException e) {
+                throw new IllegalStateException("Cannot find field (" + fd.getName() + ") on class (" + declaringClass + ").",
+                        e);
+            }
+        });
+
+        whenMetadataIsOnMethod(md -> {
+            try {
+                out[0] = declaringClass.getDeclaredMethod(md.getName()).getGenericReturnType();
+            } catch (NoSuchMethodException e) {
+                throw new IllegalStateException("Cannot find method (" + md.getName() + ") on class (" + declaringClass + ").",
+                        e);
+            }
+        });
+
+        return out[0];
```

Comment:
This code is weird.
The fact that the methods accept lambdas forces you to use the array trick to represent an effectively-final variable.
Plus this return value is quite a nasty side-effect of those lambdas.

I suggest:

- the methods return the value as opposed to storing it as a side-effect.
- the method calls turn into `if (metadataIsOnField()) then return ...`

That way, all the weirdness of this method disappears, and it will be easier to understand.